(WIP) ssx: Add mutex with checkpointing mechanism

Lazin opened the pull request:
The mutex is named and works pretty much the same way our mutex works. The difference is that every time the mutex is acquired the caller has to pass a string that contains the name of the callsite.

If the mutex is acquired the checkpoint is created. The checkpoint is simply a name of the callsite which was able to acquire the mutex and the timestamp. This information is stored in the mutex until the mutex is releasaed. When this happens the checkpoint is erased.

If any error occured the checkpoint information is included into the exception.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [x] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

Lazin committed:
ssx: Add mutex with checkpointing mechanism

The mutex is named and works pretty much the same way our mutex works.
The difference is that every time the mutex is acquired the caller has
to pass a string that contains the name of the callsite.

If the mutex is acquired the checkpoint is created. The checkpoint is
simply a name of the callsite which was able to acquire the mutex and
the timestamp. This information is stored in the mutex until the mutex
is releasaed. When this happens the checkpoint is erased.

If any error occured the checkpoint information is included into the
exception.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--bfc3ecd0-4c7d-4da7-9de7-2829bf72c077-->
<!--TEST RESULTS-->
<!--2abb2179-08a6-4089-aeca-cbf49ad4e506-->
<!--TEST RESULTS-->
<!--02b0e6c6-108f-40f3-b451-3baa6d0f4148-->
<!--TEST RESULTS-->
<!--a1079558-0481-49b9-afc2-273557b3b9c1-->
<!--TEST RESULTS-->
<!--ac3ee6c3-2e2b-4dfb-8179-5785810bde16-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66640">build#66640</a></summary>


| test_class                   | test_method                              | test_arguments                                                                                                                       | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------|:-----------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerGroupOffsetResetTest | test_stress_consumer_group_commits       |                                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66640#019722c2-2487-4652-b8ed-e460537dc03e | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RedpandaConnectIcebergTest   | test_translating_avro_serialized_records | {"cloud_storage_type": 1}                                                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66640#019722b5-768d-48e0-8cf2-307c1c81d653 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RandomNodeOperationsTest     | test_node_operations                     | {"cloud_storage_type": 2, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66640#019722c2-2485-4e6c-a590-72d22028d0ac | FLAKY         | 20/21    | upstream reliability is '97.36842105263158'. current run reliability is '95.23809523809523'. drift is 2.13033 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66650">build#66650</a></summary>


| test_class                        | test_method                           | test_arguments                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------------------|:--------------------------------------|:------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterHealthOverviewTest         | cluster_health_overview_baseline_test |                                                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66650#019725ea-872b-40cb-84ed-1e6811a92d5c | FLAKY         | 19/21    | upstream reliability is '97.26962457337885'. current run reliability is '90.47619047619048'. drift is 6.79343 and the allowed drift is set to 50. The test should PASS  |
| DatalakeBatchingTest              | test_batching                         | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "expect_large_files": true, "query_engine": "spark"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66650#019725ea-872b-40cb-84ed-1e6811a92d5c | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| PartitionForceReconfigurationTest | test_basic_reconfiguration            | {"acks": 1, "controller_snapshots": true, "restart": true}                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66650#0197260b-108a-46e1-8d13-7c0745b7c7ff | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              |
| DisablingPartitionsTest           | test_disable                          |                                                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66650#0197260b-108d-43f3-841a-afbd25a10b20 | FLAKY         | 18/21    | upstream reliability is '79.57610789980733'. current run reliability is '85.71428571428571'. drift is -6.13818 and the allowed drift is set to 50. The test should PASS |
| WriteCachingFailureInjectionTest  | test_unavoidable_data_loss            |                                                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66650#019725ea-872e-4b67-8d1e-e206ae95a079 | FLAKY         | 20/21    | upstream reliability is '99.20318725099602'. current run reliability is '95.23809523809523'. drift is 3.96509 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66655">build#66655</a></summary>


| test_class                    | test_method                               | test_arguments                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:------------------------------|:------------------------------------------|:-------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudRetentionTest            | test_cloud_retention                      | {"cloud_storage_type": 2, "max_consume_rate_mb": null} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66655#019728d5-444c-4980-9d1e-64546092f36d | FLAKY         | 20/21    | upstream reliability is '97.5'. current run reliability is '95.23809523809523'. drift is 2.2619 and the allowed drift is set to 50. The test should PASS                |
| DatalakeDelayedEnablementTest | test_enabling_iceberg_in_existing_cluster | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66655#019728d5-444c-4980-9d1e-64546092f36d | FLAKY         | 18/21    | upstream reliability is '97.19101123595506'. current run reliability is '85.71428571428571'. drift is 11.47673 and the allowed drift is set to 50. The test should PASS |
| RestCatalogConnectionTest     | test_redpanda_connection_to_rest_catalog  | {"cloud_storage_type": 1}                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66655#019728d5-444d-4095-9f36-23d4663c990b | FLAKY         | 19/21    | upstream reliability is '99.20634920634922'. current run reliability is '90.47619047619048'. drift is 8.73016 and the allowed drift is set to 50. The test should PASS  |
| SimpleEndToEndTest            | test_relaxed_acks                         | {"write_caching": false}                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66655#019728d5-444d-4095-9f36-23d4663c990b | FLAKY         | 20/21    | upstream reliability is '99.46524064171123'. current run reliability is '95.23809523809523'. drift is 4.22715 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66660">build#66660</a></summary>


| test_class                 | test_method                                       | test_arguments                                                                                                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   |
|:---------------------------|:--------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudRetentionTest         | test_cloud_retention                              | {"cloud_storage_type": 2, "max_consume_rate_mb": null}                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66660#01972bdd-af66-41a3-a9d9-cbd292250a7e | FLAKY         | 13/21    |                                                                                                                                                                          |
| PartitionReassignmentsTest | test_add_partitions_with_inprogress_reassignments |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66660#01972bdd-af67-425b-b4f6-090ed3972be8 | FLAKY         | 20/21    | upstream reliability is '88.39848675914249'. current run reliability is '95.23809523809523'. drift is -6.83961 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest   | test_node_operations                              | {"cloud_storage_type": 2, "enable_failures": false, "mixed_versions": true, "with_chunked_compaction": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66660#01972bdd-af66-41a3-a9d9-cbd292250a7e | FLAKY         | 16/18    |                                                                                                                                                                          |
| RandomNodeOperationsTest   | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66660#01972bdd-af66-4186-b57c-7ea9ac21c6f9 | FLAKY         | 20/21    | upstream reliability is '99.21875'. current run reliability is '95.23809523809523'. drift is 3.98065 and the allowed drift is set to 50. The test should PASS            |
| RandomNodeOperationsTest   | test_node_operations                              | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": true, "with_chunked_compaction": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66660#01972bdd-af66-41a3-a9d9-cbd292250a7e | FLAKY         | 14/17    |                                                                                                                                                                          |
| DisablingPartitionsTest    | test_disable                                      |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66660#01972bd9-25e9-4ebf-b819-ed8d5fadc827 | FLAKY         | 19/21    | upstream reliability is '79.57610789980733'. current run reliability is '90.47619047619048'. drift is -10.90008 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66828">build#66828</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:---------------------|:------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66828#01973b15-6bdf-482c-be96-5c6f41d6944f | FLAKY         | 20/21    | upstream reliability is '98.59154929577466'. current run reliability is '95.23809523809523'. drift is 3.35345 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": true, "with_chunked_compaction": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66828#01973b15-6bdc-4dfe-851d-4f9dbd67818d | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             |
| SaslPlainTest            | test_plain_authn     | {"client_type": 2, "sasl_plain_enabled": true, "scram_type": 2}                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66828#01973b15-6bdf-482c-be96-5c6f41d6944f | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<!--/body-->

Lazin committed:
ssx: Add execution_monitor class

The monitor is checking a set of site periodically and can detects
stalls and unexpected shutdown of the process.

The detector is using a periodic timer to check a set of pre-existing
callsties. The callsite is a simple data structure that contains a time
of the last checkpoint. There is a finite set of callsites.

The process invokes the 'checkpoint' method of the callsite while it
progresses.

There are also shutdown callsites. These are callsites which are reached
when the process is being stopped.

The execution monitor checks normal callsites and shutdown callsites
every stall-check-interval.

If all normal callsites weren't checkpointed within the stall interval
the stall is detected. If any shutdown callsite was checkpointed before
the 'expect_shutdown' was called the unexpected shutdown is detected.

In both cases the monitor invokes the callback provided by the user.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
archival: Use checkpoint_mutex in the ntp_archiver

Change _mutex and _uploads_active semaphores to be a cp-mutex. Update
acquire callsites.

The goal here is to have more visibility in a situation when something
got stuck.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
fixup! exec mon

Lazin committed:
archival: Use execution monitor in ntp_archiver

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

dotnwat reviewed (commented):
should we have a single mutex implementation that automatically has features like this, and then we can also add deadlock detection etc...?

Lazin committed:
fixup! update execution monitor

Lazin committed:
fixup! ntp-archiver

Lazin commented:
closed in favor of https://github.com/redpanda-data/redpanda/pull/26381
